plenum/test/batching_3pc/catch-up/test_clearing_requests_after_catchup.py

```diff
@@ -0,0 +1,126 @@
+import sys
+
+from plenum.test.restart.helper import restart_nodes
+from plenum.test.stasher import delay_rules
+
+from plenum.server.quorums import Quorum
+
+from stp_core.loop.eventually import eventually
+
+from plenum.test.delayers import cDelay, pDelay, ppDelay
+from plenum.test.helper import sdk_send_batches_of_random_and_check, \
+    sdk_send_batches_of_random
+
+from plenum.test.checkpoints.conftest import chkFreqPatched, reqs_for_checkpoint
+
+CHK_FREQ = 5
+LOG_SIZE = 3 * CHK_FREQ
+
+req_num = CHK_FREQ * 4
+howlong = 100
+ledger_id = 1
+another_key = 'request_2'
+
+
+def node_caughtup(node, old_count):
+    assert node.spylog.count(node.allLedgersCaughtUp) > old_count
+
+
+def test_clearing_forwarded_preprepared_request(
+        looper, chkFreqPatched, reqs_for_checkpoint, txnPoolNodeSet,
+        sdk_pool_handle, sdk_wallet_steward):
+    # Case when backup ordered correctly, but primary had problems.
+    # As a result, master will execute caughtup txns and will be removed
+    # from requests queues
+    behind_node = txnPoolNodeSet[-1]
+
+    sdk_send_batches_of_random_and_check(looper, txnPoolNodeSet, sdk_pool_handle,
+                                         sdk_wallet_steward, CHK_FREQ, CHK_FREQ)
+    with delay_rules(behind_node.nodeIbStasher,
+                     pDelay(delay=sys.maxsize, instId=0),
+                     cDelay(delay=sys.maxsize, instId=0)):
+        count = behind_node.spylog.count(behind_node.allLedgersCaughtUp)
+
+        sdk_send_batches_of_random(looper, txnPoolNodeSet, sdk_pool_handle,
+                                   sdk_wallet_steward, req_num, req_num)
+
+        looper.run(eventually(node_caughtup, behind_node, count, retryWait=1))
+
+    assert all(r.executed and r.forwardedTo == 0 for r in behind_node.requests.values())
```

Comment:
Why can't we check that requests queue (as well as queue in re_authenticator and all replicas) is empty?

Reply:
Fixed